### PR TITLE
Automated Transfer Validation API

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Adds new API for automated SDK validation of checksums on transfers.

https://github.com/Azure/azure-sdk-for-net/pull/27259 removed this feature from GA release. This reenables the feature. Code review can be kept to netstandard2.0 files and clients themselves, as the core code has already gone through thorough review.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 1]
```diff
@@ -279,6 +288,7 @@ public partial class AppendBlobOpenWriteOptions
         public long? BufferSize { get { throw null; } set { } }
         public Azure.Storage.Blobs.Models.AppendBlobRequestConditions OpenConditions { get { throw null; } set { } }
         public System.IProgress<long> ProgressHandler { get { throw null; } set { } }
+        public Azure.Storage.UploadTransferValidationOptions TransferValidationOptions { get { throw null; } set { } }
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:291]: I also see `UploadTransferValidationOptions` as a property on `BlobClientOptions`.  Are the `*ValidationTransferOptions` client-level parameters, API-level parameters, or both?

File: sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs [hunk 2]
```diff
@@ -1286,9 +1251,8 @@ public new AppendBlobClient WithEncryptionScope(string encryptionScope)
                         response = await AppendBlobRestClient.AppendBlockAsync(
                             contentLength: (content.Length - content.Position),
                             body: content,
-                            // TODO #27253
-                            //transactionalContentCrc64: hashResult?.StorageCrc64,
-                            transactionalContentMD5: transactionalContentMD5, //hashResult?.MD5,
+                            transactionalContentCrc64: hashResult?.StorageCrc64,
+                            transactionalContentMD5: hashResult?.MD5,
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:1255]: Any issue if both `transactionalContentCrc64` and `transactionalContentMD5` are set?

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 3]
```diff
@@ -937,28 +941,10 @@ private void SetNameFieldsIfNull()
...
-            //        //        Algorithm = TransactionalHashAlgorithm.MD5
-            //        //    }
-            //        //    : default,
-            //    };
-            //}
             Response<BlobDownloadStreamingResult> response = await DownloadStreamingDirect(
                 range,
                 conditions,
-                rangeGetContentHash,
+                rangeGetContentHash.ToValidationOptions(),
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:947]: Does `ToValidationOptions()` handle null?  (Not sure if this matters)

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlockBlobRestClient.cs [hunk 4]
```diff
@@ -40,7 +40,7 @@ public BlockBlobRestClient(ClientDiagnostics clientDiagnostics, HttpPipeline pip
...
+        internal HttpMessage CreateUploadRequest(long contentLength, Stream body, int? timeout, byte[] transactionalContentMD5, byte[] transactionalContentCrc64, string blobContentType, string blobContentEncoding, string blobContentLanguage, byte[] blobContentMD5, string blobCacheControl, IDictionary<string, string> metadata, string leaseId, string blobContentDisposition, string encryptionKey, string encryptionKeySha256, EncryptionAlgorithmTypeInternal? encryptionAlgorithm, string encryptionScope, AccessTier? tier, DateTimeOffset? ifModifiedSince, DateTimeOffset? ifUnmodifiedSince, string ifMatch, string ifNoneMatch, string ifTags, string blobTagsString, DateTimeOffset? immutabilityPolicyExpiry, BlobImmutabilityPolicyMode? immutabilityPolicyMode, bool? legalHold)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlockBlobRestClient.cs:43]: TODO update https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs/src/autorest.md and regenerate after https://github.com/Azure/azure-rest-api-specs/pull/19749 merges

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobDownloadOptions.cs [hunk 5]
```diff
@@ -1,87 +1,88 @@
...
+        /// Storage size limits for requesting a transactional hash. See the
+        /// <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob">
+        /// REST documentation</a> for range limitation details.
+        /// </summary>
+        public DownloadTransferValidationOptions TransferValidationOptions { get; set; }
 
+        /// <summary>
+        /// Default constructor.
+        /// </summary>
+        public BlobDownloadOptions()
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/BlobDownloadOptions.cs:49]: Nit: I don't think we need this constructor.

File: sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs [hunk 6]
```diff
@@ -1195,129 +1177,112 @@ public new PageBlobClient WithEncryptionScope(string encryptionScope)
...
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{PageInfo}"/> describing the
+        /// state of the updated pages.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Response<PageInfo> UploadPages(
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:1224]: I think the optional parameters here need `= default`.

File: sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs [hunk 7]
```diff
@@ -1195,129 +1177,112 @@ public new PageBlobClient WithEncryptionScope(string encryptionScope)
...
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{PageInfo}"/> describing the
+        /// state of the updated pages.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual async Task<Response<PageInfo>> UploadPagesAsync(
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:1272]: I think the optional parameters here need `= default`.

File: sdk/storage/Azure.Storage.Common/src/Shared/TransferValidationOptionsExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Storage
+{
+    internal static class TransferValidationOptionsExtensions
+    {
+        public static ValidationAlgorithm ResolveAuto(this ValidationAlgorithm validationAlgorithm)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/TransferValidationOptionsExtensions.cs:8]: Do we need to worry about the null case?

File: sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs (hunk 2 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:1255]: Service will reject the request. However, a caller can only set one, and the code that calculates this object will respect that.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 3 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:947]: Yes

```csharp
public static UploadTransferValidationOptions ToValidationOptions(this byte[] md5)
    => md5 == default
        ? default
        : new UploadTransferValidationOptions
        {
            Algorithm = ValidationAlgorithm.MD5,
            PrecalculatedChecksum = md5
        };
```

File: sdk/storage/Azure.Storage.Common/src/Shared/TransferValidationOptionsExtensions.cs (hunk 8 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/TransferValidationOptionsExtensions.cs:8]: Can't be null. `ValidationAlgorithm` is an enum.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:291]: What happens if the API-level options and client-level options conflict?  Which takes precedence?

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeExtensions.cs [hunk 9]
```diff
@@ -504,43 +504,38 @@ internal static BlobOpenReadOptions ToBlobOpenReadOptions(this DataLakeOpenReadO
...
-        //    };
-        //}
+        internal static BlobDownloadOptions ToBlobBaseDownloadOptions(this DataLakeFileReadOptions options)
+        {
+            if (options == null)
+            {
+                return null;
+            }
+
+            return new BlobDownloadOptions()
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeExtensions.cs:518]: Looks like we have a merge conflict here

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 10]
```diff
@@ -2819,12 +2795,11 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
-        //[EditorBrowsable(EditorBrowsableState.Never)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
         public virtual Response<FileDownloadInfo> Read(
-            HttpRange range = default,
-            DataLakeRequestConditions conditions = default,
-            bool rangeGetContentHash = default,
+            HttpRange range,
+            DataLakeRequestConditions conditions,
+            bool rangeGetContentHash,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:2803]: Remove `= default` and add the suppression to this overload.  Here's an example - https://github.com/Azure/azure-sdk-for-net/blob/2fa1dc608acf4b051c175b672b5363eaa668870f/sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs#L1002

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 11]
```diff
@@ -2892,12 +2867,11 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
-        //[EditorBrowsable(EditorBrowsableState.Never)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
         public virtual async Task<Response<FileDownloadInfo>> ReadAsync(
-            HttpRange range = default,
-            DataLakeRequestConditions conditions = default,
-            bool rangeGetContentHash = default,
+            HttpRange range,
+            DataLakeRequestConditions conditions,
+            bool rangeGetContentHash,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:2875]: Same as previous comment.

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 12]
```diff
@@ -3266,15 +3238,14 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual Response ReadTo(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
             Stream destination,
-            DataLakeRequestConditions conditions = default,
+            DataLakeRequestConditions conditions,
             //IProgress<long> progressHandler,
-            StorageTransferOptions transferOptions = default,
+            StorageTransferOptions transferOptions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:3249]: Same as previous comment.

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 13]
```diff
@@ -3330,15 +3301,14 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual Response ReadTo(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
             string path,
-            DataLakeRequestConditions conditions = default,
+            DataLakeRequestConditions conditions,
             //IProgress<long> progressHandler,
-            StorageTransferOptions transferOptions = default,
+            StorageTransferOptions transferOptions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:3312]: Same

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 14]
```diff
@@ -3394,15 +3364,14 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual async Task<Response> ReadToAsync(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
             Stream destination,
-            DataLakeRequestConditions conditions = default,
+            DataLakeRequestConditions conditions,
             //IProgress<long> progressHandler,
-            StorageTransferOptions transferOptions = default,
+            StorageTransferOptions transferOptions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:3375]: Same

File: sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs [hunk 15]
```diff
@@ -3459,15 +3428,14 @@ Response<Blobs.Models.BlobDownloadStreamingResult> response
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual async Task<Response> ReadToAsync(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
             string path,
-            DataLakeRequestConditions conditions = default,
+            DataLakeRequestConditions conditions,
             //IProgress<long> progressHandler,
-            StorageTransferOptions transferOptions = default,
+            StorageTransferOptions transferOptions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs:3439]: Same

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 16]
```diff
@@ -1945,32 +1946,20 @@ private void SetNameFieldsIfNull()
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual Response<ShareFileDownloadInfo> Download(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
-            HttpRange range = default,
-            bool rangeGetContentHash = default,
-            ShareFileRequestConditions conditions = default,
+            HttpRange range,
+            bool rangeGetContentHash,
+            ShareFileRequestConditions conditions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:1956]: Same

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs [hunk 17]
```diff
@@ -2077,33 +2055,20 @@ private void SetNameFieldsIfNull()
...
 #pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual async Task<Response<ShareFileDownloadInfo>> DownloadAsync(
 #pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
-            HttpRange range = default,
-            bool rangeGetContentHash = default,
-            ShareFileRequestConditions conditions = default,
+            HttpRange range,
+            bool rangeGetContentHash,
+            ShareFileRequestConditions conditions,
             CancellationToken cancellationToken = default)
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs:2065]: Same

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobDownloadOptions.cs (hunk 5 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/BlobDownloadOptions.cs]: Hmm. There is a private copy constructor in place but I'm not sure what it was doing back when it was made. Will investigate.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 1 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:291]: method arg when not null takes precedence. An instance can be constructed with `None` as a disable.

**Review by seanmcc-msft — APPROVED** [review]
Approved pending CI and changelogs

**Pull request merged**